src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+            let end_poff = Ao.end_poff ao in
+            (off := Int63.Syntax.(!off + end_poff));
+            { idx = chunk_idx; start_off; ao }
+      in
+      try Ok (v chunk_num create_chunk)
+      with OpenInventoryError err ->
+        Error (err : open_error :> [> open_error ])
+  end
+
+  type t = { inventory : Inventory.t }
+
+  (* A lightweight wrapper around creating/opening append only files as chunks *)
+  module Ao_chunk = struct
+    let path = Layout.V4.suffix_chunk
+
+    let create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure =
+      let path = path ~root ~chunk_idx in
+      Ao.create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_procedure
+
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
+      let path = path ~root ~chunk_idx in
+      Ao.open_rw ~path ~end_poff ~dead_header_size ~auto_flush_threshold
+        ~auto_flush_procedure
+
+    let open_ro ~root ~chunk_idx ~end_poff ~dead_header_size ~is_appendable =
+      let open Result_syntax in
+      let path = path ~root ~chunk_idx in
+      (* Passing [end_poff] is required by Ao for its consistency check (and
+         only applies to the appendable chunk).
+         TODO: We could change its API in the future so we can not do
+         [Io.size_of_path], which is admittedly very silly.*)
+      let* end_poff =
+        if is_appendable then Ok end_poff else Io.size_of_path path
+      in
+      Ao.open_ro ~path ~end_poff ~dead_header_size
+  end
+
+  let create_rw ~root ~start_idx ~overwrite ~auto_flush_threshold
+      ~auto_flush_procedure =
+    let open Result_syntax in
+    let chunk_idx = start_idx in
+    let+ ao =
+      Ao_chunk.create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure
+    in
+    let chunk = { idx = chunk_idx; start_off = Int63.zero; ao } in
+    let inventory = Inventory.v 1 (Fun.const chunk) in
+    { inventory }
+
+  let open_rw ~root ~end_poff ~start_idx ~chunk_num ~dead_header_size
+      ~auto_flush_threshold ~auto_flush_procedure =
+    let open Result_syntax in
+    let open_chunk ~chunk_idx ~is_appendable =
+      match is_appendable with
+      | true ->
+          Ao_chunk.open_rw ~root ~chunk_idx ~end_poff ~auto_flush_threshold
+            ~auto_flush_procedure ~dead_header_size
+      | false ->
+          Ao_chunk.open_ro ~root ~chunk_idx ~end_poff ~dead_header_size
+            ~is_appendable
+    in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let open_ro ~root ~end_poff ~dead_header_size ~start_idx ~chunk_num =
+    let open Result_syntax in
+    let open_chunk = Ao_chunk.open_ro ~root ~end_poff ~dead_header_size in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
```

Comment:
I think this is wrong -  the `end_poff` is used upstreamed as the end poff of the whole suffix abstraction, but here it returns only the end poff of the last chunk. I should be the sum over all chunks.

Reply:
Thanks for bringing this up. I think this is worth some discussion.

The way I understand how the suffix's `poff` is currently used in the control file is as a control mechanism to know/verify what data we have written to disk (as is seen in the append only file's consistency check). In this sense, the current code is correct. It tracks the physical offset of the appendable chunk for consistency checks. No other chunk can be changed so its `end_poff` is necessarily equivalent to its `Io.size_of_path` (the awkward code in `Ao_chunk.open_ro` that does exactly this).

Summing the values would give the chunked suffix's length and final offset but doesn't really correlate with a physical offset

While working on this I intentionally wanted to keep the existing code working with minimal extra changes but I do think we need to change this. Here is my proposal for a future PR:

- rename control file field to `suffix_consistency_poff` (or something else -- open to ideas!)
- rename `end_poff` in chunked file to `consistency_poff`
- move consistency checking up from append only to chunked suffix and remove `end_poff` from append only's api

Reply:
The dispatcher needs to know the end offset for the suffix file in order to verify that a read is within bounds.  This does not correspond to a physical offset, indeed. It's "a virtual offset corresponding to the physical end offset, if the suffix was a single file". I'll call it `suffix_end_off`.

We can either (I am reiterating what you say):
- keep `end_poff` as is, keep the control file as is (tracking the `end_poff` of the last chunk), but then we need a function to compute the  `suffix_end_off` - solution works for me, I do find the concept of `suffix_end_off` a bit weird. 
- replace the `end_poff` in the chunked file and the control file with the `suffix_end_off` - I'm not sure if this is what you are proposing for a future PR?

Reply:
Ah, I just looked in the dispatcher and see its usage of `Suffix.end_poff` -- this does need to be updated to be the suffix end offset. Before chunks, this makes sense as an `end_poff`, but there is no "end poff" for a chunked suffix. I don't necessarily love adding another virutalized set of offsets, but the suffix's offset range will necessarily be virtualized for chunks since it doesn't map directly to a single range of physical offsets any longer. And it is neatly contained within `Chunked_suffix` which helps reduce the scope of offsets to consider.

Thanks for pointing out dispatcher's use of the end offset for the suffix. I think the following proposal addresses everything:

- Add `end_off` to `Chunked_suffix` to represent the last offset of the suffix. This can be a sum of the chunks poffs but I don't think tracking it in the control file is needed.
- Rename `Chunked_suffix.end_poff` to `consistency_poff` (or something else -- but I think "end poff" doesn't make sense in the context of a chunked suffix) and also the control file field to reflect the name change. This is the offset we need to track in the control file for consistency checks since we just need to check the appendable chunk's poff upon open. As a part of this, also move the consistency check from append only to chunked suffix (like mentioned previously).

I will make the first change in this PR since you rightly pointed out a correctness issue. The second one can be a future PR since it is mostly a cosmetic name change, but it will also get rid of some of the awkward `end_poff` code (like in `Ao_chunk.open_ro`)  which will be nice.

Reply:
Aha, I missed that. I'll just review in your PR since you have already made the change. :+1: